search: Add title to the search bar when in "sender:" narrow & rewrite is_common_narrow.

palashb01 opened the pull request:
- Display title to the search bar in `sender:` narrow:

    -   for me: Messages sent by you
    -   for any user: Messages sent by User

- Rewrite the `is_common_narrow`
---------------------

Fixes: #18690
Czo: [is_common_narrow topic](https://chat.zulip.org/#narrow/stream/6-frontend/topic/is_common_narrow.20rewrite)

--------------

**Screenshots and screen captures:**

<details>
<summary>Before:</summary>
<br>


<img width="890" alt="Screenshot 2023-08-13 at 1 16 56 AM" src="https://github.com/zulip/zulip/assets/66828942/4a514dc3-cde2-4117-9c6b-1d80464f2173">




</details>

<details>
<summary>After:</summary>
<br>


<img width="890" alt="Screenshot 2023-08-13 at 1 12 08 AM" src="https://github.com/zulip/zulip/assets/66828942/27cbc540-2279-4157-975d-240f693c2d3a">

-------------------

<img width="890" alt="Screenshot 2023-08-13 at 1 12 14 AM" src="https://github.com/zulip/zulip/assets/66828942/baec1635-6724-47e8-9380-038d319f549e">

</details>

-------------

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

palashb01 commented:
@laurynmm , could you please review the PR, and suggest if any changes are required, thanks :)

laurynmm commented on web/src/filter.js:
```diff
@@ -550,29 +550,57 @@ export class Filter {
+        // (which are_all_conversation_messages_of_a_message_in_view does not check as starred messages are always read)
         // is: mentioned
-        // (for which can_mark_messages_read returns false, since they don't have conversation context)
+        // (for which are_all_conversation_messages_of_a_message_in_view returns false, since they don't have conversation context)
```
I feel like writing a better comment here would be helpful with this refactor too.

palashb01 replied:
Updated the comment, do let me know if you want me to make anymore changes,

laurynmm reviewed (commented):
@palashb01 - Thanks for picking up the work on making these changes! I had a few comments on how we can make these changes more complete re: updating code comments and making sure that the "sender" operator behaves like a common narrow. Let me know if you have any questions!

palashb01 commented on web/src/filter.js:
```diff
@@ -550,29 +550,57 @@ export class Filter {
 
+    // This function is used to check whether all the messages of the conversation
+    // are present or not of which a message is present in the view.
+    are_all_conversation_messages_of_a_message_in_view() {
```
Updated your suggested name and comment, also started the discussion on the CZO : [Topic](https://chat.zulip.org/#narrow/stream/6-frontend/topic/is_common_narrow.20rewrite/near/1627628)

palashb01 commented on web/src/filter.js:
```diff
@@ -579,6 +579,10 @@ export class Filter {
 
+        if (_.isEqual(term_types, ["sender"]) && people.get_by_email(this.operands("sender")[0])) {
+            return true;
+        }
```
Updated and thanks for the suggestion and insights.

palashb01 commented on web/src/filter.js:
```diff
@@ -750,6 +755,20 @@ export class Filter {
+                    },
+                );
+            }
+        }
```
updated the `narrow.js` similar to `dm`, thanks for the quick catch of the extra work

palashb01 commented on web/src/filter.js:
```diff
@@ -593,6 +597,7 @@ export class Filter {
         // stream, stream + topic,
         // is:dm, dm,
         // is:resolved
+        // sender
```
> Let's work to update this comment in the prep commit so that we don't list out all the filters here.

I am not sure what you mean over here, could you please elaborate.

> Also, the comment above is_common_narrow with the link to the paper document will become even more out of date with these changes, so we should figure out how we want to deal with that. We could either update that document or update the comment here to better describe what a common narrow is. Maybe worth bringing up in the CZO conversation about this refactor.

Started the discussion: [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/is_common_narrow.20rewrite/near/1627630)

palashb01 commented:
@laurynmm , thanks for the review, I have updated the PR, could you please review again, thanks :)

laurynmm reviewed (commented):
@palashb01 - Most of the updates look good, I had a few comments/questions, let me know if you need anything clarified.

A few general things...

**Code comments**: Let's really do a good job of cleaning these up here. If the comment is just describing what the code does, let's remove it. If the comment is out of date either currently or after these changes, let's remove it or make it accurate.

**What each function does**: Your first commit message wasn't clear, so let's make sure we're on the same page about each function's purpose...
- the new helper function: checks a few, basic filters that are true for both functions listed below, and the general rule here is that these filters return all messages of a conversation
- `is_common_narrow`: checks for filters that have a special message view header, which also means they have special behavior when exiting search
- `calc_can_mark_as_read`: this is the non-conversation view check for whether messages can be marked read in a message list view with this filter

**Naming thoughts**: The helper function does not support combining valid filter terms (unlike `supports_collapsing_recipients`), so maybe that's where we focus for the semantic name. The helper function only supports 1 term filters, with two exceptions: stream/topic and the empty "All view".

palashb01 commented on web/src/filter.js:
```diff
@@ -501,6 +501,26 @@ export class Filter {
+        const term_types = this.sorted_term_types();
+        if (_.isEqual(term_types, [])) {
+            // All view
+            return true;
```
// All view not needed inside is_common_narrow because all messages narrow has a filter as undefined and therefore we are already handling that case everywhere.

laurynmm replied:
Okay. I see that you've worked on explaining this in the first commit message. But that commit message is very difficult to read and understand as it is. I suggest rewording it for clarity and simplicity.

Here's a go at a revision from me:

```console
filter: Create helper function that checks basic filter terms.

Currently, `is_common_narrow` relies on `calc_can_mark_messages_read`
to check a number of filters that are "common narrows" (which means
they have a special message header title and different behavior when
exiting a search view).

Creates a helper function, `all_messages_of_conversation_included`,
that is used in both the above functions.

The check for an empty array of term types is unnecessary for
`is_common_narrow` because the "All messages" view has an undefined
filter and the "in-home" term type used in the "zhome" message list
is covered. So that empty array check is not moved to the helper
function, and instead stays only in `calc_can_mark_messages_read`.

The helper function checks for a single term type (except in the case
of topic, which is checked in combination with stream) that will return
all the messages of a conversation.
```

palashb01 replied:
Done

palashb01 commented on web/src/filter.js:
```diff
@@ -501,6 +501,26 @@ export class Filter {
+    all_messages_of_conversation_included() {
         // Arguably this should match supports_collapsing_recipients.
         // We may want to standardize on that in the future.  (At
         // present, this function does not allow combining valid filters).
```
Changed it.

palashb01 commented on web/src/filter.js:
```diff
@@ -531,48 +551,31 @@ export class Filter {
+        // all_messages_of_conversation_included tests the following filters:
         // stream, stream + topic,
         // is:dm, dm,
         // is:resolved
```
Got it, removed it thanks for the suggestion

palashb01 commented on web/src/filter.js:
```diff
@@ -531,48 +551,31 @@ export class Filter {
+        // (which all_messages_of_conversation_included does not check as starred messages are always read)
         // is: mentioned
-        // (for which can_mark_messages_read returns false, since they don't have conversation context)
+        // (for which all_messages_of_conversation_included returns false, since they don't have conversation context)
```
Removed them.

palashb01 commented on web/src/filter.js:
```diff
@@ -725,6 +730,23 @@ export class Filter {
         }
+        if (term_types.length === 1 && _.isEqual(term_types, ["sender"])) {
+            const email = this.operands("sender")[0];
+            const user = people.get_by_email(email);
```
So the behaviour of this and the `dm` filter are both the same If we type invalid user then it shows, `This user does not exist!`, do you want me to make any changes ?


<img width="767" alt="Screenshot 2023-08-27 at 1 42 00 AM" src="https://github.com/zulip/zulip/assets/66828942/d23ca2e8-a120-4872-ab66-6bf2349b4969">

laurynmm replied:
The warnings I was seeing are from the `dm` view with an unknown email/user, so this isn't an issue for this PR.

palashb01 commented:
@laurynmm , thank you very much for the review and the suggestion. I have modified the commit description to more clearly convey the reasoning, and I have also implemented the changes you recommended. Regarding the naming issue, I have posted a message in the CZO thread to gather suggestions. Could you please take another look at the PR? Thank you :)

palashb01 commented:
@laurynmm I have updated the name and changes, as suggested by you and Tim, could you please review the PR, thanks :)

laurynmm reviewed (commented):
@palashb01 - This is looking good! I had a few more small inline notes for a few more clean ups we can do in the prep commit. The implementation of `sender` as a common narrow in the second commit looks good.

One more, not inline note for a comment that needs to be cleaned up:
- The `TODO` comment about marking messages as read in the helper function can be deleted. We now have a user-level setting that restricts marking messages as read to conversation (aka non-interleaved views).

Also, once these changes are merged, I'll put updating the paper doc about common narrows on my to-do list, becuase it will be pretty out of date by then.

palashb01 commented on web/src/filter.js:
```diff
@@ -526,48 +547,22 @@ export class Filter {
-
         if (term_types.length === 1 && ["in-home", "in-all"].includes(term_types[0])) {
             return true;
         }
```
Sure, thanks for suggesting

palashb01 commented:
@laurynmm , I have updated the PR as per your suggestions, could you please review the PR, Thanks :)
I am not sure why the backend tests are failing; it seems that the setup jobs for the tests have not been successful.

laurynmm commented on web/src/filter.js:
```diff
@@ -499,6 +499,27 @@ export class Filter {
+    single_term_type_returns_all_messages_of_conversation() {
         const term_types = this.sorted_term_types();
 
         if (_.isEqual(term_types, ["stream", "topic"])) {
```
Updates look good to me.

laurynmm commented:
@palashb01 - The most recent updates look good to me. I pushed an updated version with an empty line between the new helper function and the comment for `is_common_narrow`, which triggered a rerun of the CI. Marking for integration review.

timabbott commented on web/src/filter.js:
```diff
@@ -499,8 +499,31 @@ export class Filter {
+        }
+        const term_types = this.sorted_term_types();
+        if (_.isEqual(term_types, [])) {
+            // "All messages" view
```
This comment was just "All view", which was unclear; fixed it for merging.

timabbott commented:
@palashb01 this looks great, marked to merge once CI passes.

I have a few follow-up thoughts, probably best discussed on separate chat.zulip.org threads (please start them!):
* Should we rename `is_common_narrow` to `has_title`? It's only used in one place, and I think the meaning is just that we can call `filter.get_title` and get something useful? Or maybe I could imagine getting rid of it in favor of just calling `get_title` instead to answer that question... or am I missing something about how it is different?
* I am not sure I like the "Sent by user@example.com" framing; but probably this is a thing we can defer fixing until we're doing user pills, which would just eliminate the case.
